src/openforms/submissions/tests/test_models.py

```diff
@@ -610,3 +610,10 @@ def test_get_cosigner_email(self):
 
         with self.subTest("nested"):
             self.assertEqual(submission3.cosigner_email, "test@test.nl")
+
+    def test_clear_execution_state_without_execution_state(self):
+        submission = SubmissionFactory.create()
+
+        self.assertFalse(hasattr(submission, "_execution_state"))
+
+        submission.clear_execution_state()
```

Comment:
As from the previous comment, the situation is a lot hairier.

The notion of staleness on the `SubmissionState` is broken too: the `load_execution_state` has a `refresh` and disregards references to the current state object it has returned previously and hands out a new one. That's unsafe. It probably hasn't resulted in problems, because most callers, get one, do some work and let it go out of scope and not keep it around.  I haven't checked. (I'm not a Rust compiler :wink:)

Reply:
Can we create a new issue/big topic for this? This state (submission but also variables) has grown way beyond the initial scope/ideas and I think it's definitely up for refactoring. We should now have a clearer picture of what it needs to do and should probably be able to break this out into a proper service layer.

It's important to decouple the usage throughout the stack from the backing database models, but we also need to implement the DB aspect properly for performance reasons. We should be able to collectively come up with a set of (functional) requirements. I think the test suite is implementation detail agnostic enough to be able to achieve this.